File: Minidown/Minidown/Stats.swift
// minidown — a minimal, distraction-free Markdown writer for macOS.
// Copyright (C) 2026 Asit Khanda
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the Free
// Software Foundation, either version 3 of the License, or (at your option) any
// later version. See <https://www.gnu.org/licenses/>.
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import Foundation

enum StatsMode: String, CaseIterable {
    case words
    case characters
    case reading

    var next: StatsMode {
        switch self {
        case .words: return .characters
        case .characters: return .reading
        case .reading: return .words
        }
    }

    func format(_ text: String) -> String {
        switch self {
        case .words:
            let count = wordCount(text)
            return count == 1 ? "1 word" : "\(count) words"
        case .characters:
            let count = text.count
            return count == 1 ? "1 character" : "\(count) characters"
        case .reading:
            let minutes = max(1, Int(ceil(Double(wordCount(text)) / 200.0)))
            return minutes == 1 ? "1 min read" : "\(minutes) min read"
        }
    }

    private func wordCount(_ text: String) -> Int {
        text.split { $0.isWhitespace || $0.isNewline }.count
    }
}

File: src/stats.ts
export type StatsMode = "words" | "chars" | "time";

const MODES: StatsMode[] = ["words", "chars", "time"];
const WORDS_PER_MINUTE = 220;

export function nextStatsMode(mode: StatsMode): StatsMode {
  return MODES[(MODES.indexOf(mode) + 1) % MODES.length];
}

export function countWords(text: string): number {
  return text.match(/\S+/g)?.length ?? 0;
}

export function formatStats(text: string, mode: StatsMode): string {
  switch (mode) {
    case "words": {
      const words = countWords(text);
      return `${words.toLocaleString()} ${words === 1 ? "word" : "words"}`;
    }
    case "chars": {
      const chars = text.length;
      return `${chars.toLocaleString()} ${chars === 1 ? "character" : "characters"}`;
    }
    case "time": {
      const words = countWords(text);
      if (words === 0) return "0 min read";
      const minutes = Math.ceil(words / WORDS_PER_MINUTE);
      return `${minutes} min read`;
    }
  }
}
